eggtimer.js

```diff
@@ -1,326 +1,681 @@
-  HANDLER(req, res, function (err) {
-    res.statusCode = 404
-    res.end('no such location')
-  });
-}).listen(CONFIG.port);
+const Config = JSON.parse(fs.readFileSync('config.js'));
+const WebhookHandler = createHandler({ path: Config.github_webhook_path, secret: Config.github_webhook_secret });
+const Github = new nodeGithub({ version: "3.0.0" });
+const GithubAuthentication = { type: 'token', username: Config.github_username, token: Config.github_token };
+
+const MergeFailedLabel = "S-merge-failed";
+const MergedLabel = "S-merged";
+const MergingTag = "T-merging-PR";
+const TagRegex = /(refs\/tags\/.*-PR)(\d+)$/;
+
+let PRList = [];
+let currentContext = null;
 
-HANDLER.on('error', function (err) {
+WebhookHandler.on('error', (err) => {
   console.error('Error:', err.message);
 });
 
 // https://developer.github.com/v3/activity/events/types/#pullrequestreviewevent
-HANDLER.on('pull_request_review', function(event) {
-    const url = event.payload.pull_request.url;
-    const head_sha = event.payload.pull_request.head.sha;
-    const ref = event.payload.pull_request.head.ref;
-    console.log(url + " -> pull_request_review");
-    ensurePr(url, head_sha);
-    prs[url].ref = ref;
-    populateMergeable(url);
-    populateReviews(url);
-    mergeIfReady(url);
+WebhookHandler.on('pull_request_review', (ev) => {
+    const pr = ev.payload.pull_request;
+    const review = ev.payload.review;
+    console.log("pull_request_review event:", ev.payload.id, pr.number, pr.head.sha, pr.state, review.state);
+    processEvent();
 });
 
 // https://developer.github.com/v3/activity/events/types/#pullrequestevent
-HANDLER.on('pull_request', function(event) {
-    const url = event.payload.pull_request.url;
-    const head_sha = event.payload.pull_request.head.sha;
-    const ref = event.payload.pull_request.head.ref;
-    console.log(url + " -> pull_request");
-    ensurePr(url, head_sha);
-    prs[url].ref = ref;
-    populateMergeable(url);
-    populateReviews(url);
-    mergeIfReady(url);
+WebhookHandler.on('pull_request', (ev) => {
+    const pr = ev.payload.pull_request;
+    console.log("pull_request event:", ev.payload.id, pr.number, pr.head.sha, pr.state);
+    processEvent();
 });
 
 // https://developer.github.com/v3/activity/events/types/#statusevent
-HANDLER.on('status', function(event) {
-    const sha = event.payload.sha;
-    const context = event.payload.context;
-    const state = event.payload.state;
-    var success = false;
-    switch (state) {
-        case 'success':
-            success = true;
-            break;
-        case 'pending':
-        case 'failure':
-        case 'error':
-            // success = false, still
-            break;
-        default:
-            console.error("Unknown check state '" + state + "'. success = false");
-            break;
+WebhookHandler.on('status', (ev) => {
+    const e = ev.payload;
+    console.log("status event:", e.id, e.sha, e.context, e.state);
+    if (merging(e.sha)) {
```

Comment:
No, that is not an optimization. This handles status events, coming from auto_branch tests.

Reply:
If this is not an optimization, then this is a bug: The bot should not rely on receiving any specific notification (because, for example, the bot may not be running when GitHub sends the notification).

Reply:
I don't see where exactly this bug is: we do not rely _only_ on this information. The situation when the bot was off while such events came is handled separately.

Reply:
> I don't see where exactly this bug is: we do not rely _only_ on this information.

The question is _not_ whether we rely on some other information as well. The question is whether we rely on (i.e., require) _this specific event_ to function correctly.

* If we rely on it, then it is a bug because it violates the "do not rely on receiving any specific notification" invariant.
* If we do not rely on it, then it is just an optimization (which may be safely removed to simplify the code; for now).

Are you implying that "do not rely on receiving any specific notification" is a wrong invariant to uphold? I do not think you do. Given the fact that GitHub implementation itself and bot access to the internet are not 100% reliable, I do not see why that invariant is wrong...

You have attacked an example where that invariant is important (a bot server crash) because that example is irrelevant to this use case, but that was just the simplest example to illustrate the invariant. In our deliberations, we should focus on the invariant itself.